File: 783 - Minimum Distance Between BST Nodes.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    minval = -1
    prenode = -1
    def minDiffInBST(self, root: TreeNode) -> int:
        self.traverse(root)
        return self.minval
    def traverse(self, root):
        if root is None:
            return
        self.traverse(root.left)
        if self.prenode == -1:
            self.prenode = root.val
        else:
            diff = root.val - self.prenode
            self.prenode = root.val
            if self.minval == -1:
                self.minval = diff
            else:
                self.minval = min(self.minval, diff)
        self.traverse(root.right)

File: 783 - Minimum Distance Between BST Nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
let mindiff = -1, preval = -1;
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDiffInBST = function(root) {
    mindiff = preval = -1;
    traverse(root);
    return mindiff;
};

/**
 * 
 * @param {TreeNode} root 
 * @return {void}
 */

function traverse(root) {
    if (root == null) return;
    traverse(root.left);
    if (preval == -1) {
        preval = root.val;
    } else {
        let diff = root.val - preval;
        preval = root.val;
        if (mindiff == -1) {
            mindiff = diff;
        } else {
            mindiff = mindiff < diff ? mindiff : diff;
        }
    }
    traverse(root.right);
}